VerilogForVerificationStyle.md

```diff
@@ -0,0 +1,826 @@
+| Agent                             | `<dut>_agent m_<dut>_agent;`                 |
+| Agent level config object         | `rand <dut>_agent_cfg m_<dut>_agent_cfg;`    |
+| Agent coverage collection object  | `<dut>_agent_cov cov;`                       |
+| Driver                            | `<dut>_driver driver;`                       |
+| Monitor                           | `<dut>_monitor monitor;`                     |
+| Scoreboard                        | `<dut>_scoreboard scoreboard;`               |
+| Virtual sequencer                 | `<dut>_virtual_sequencer virtual_sequencer;` |
+| Sequencer                         | `<dut>_sequencer <dut>_sequencer;`           |
+
+The UVM testbench generation script will automatically create skeletons of these
+components with this naming scheme, so almost all of this naming is automated
+out.
+Additional naming guidelines:
+
+1. When instantiating any objects, the `name` argument to `type_id::create()`
+must match the name of the variable being assigned to.
+This ensures that any reporting messages refer to the right objects.
+2. To avoid name collisions, it is recommended that global types and types defined
+within a package have unique prefixes. This includes typedefs, enumerated
+types/values, package parameters, package localparams, any functions/tasks, and
+macros. Note that all package variables (parameters, etc.) must have a specified type.
+3. When instantiating a user-defined class member variable that is a class object:
+    *   Prefix the instance name of the class object with `m_`, as such:
+
+```systemverilog
+my_animal m_my_animal;
+```
+
+This scheme does not apply to any class member variables that are
+not themselves class objects.
+
+* If multiple instantiations of the class are desired without an array
+to store them, use `<class_name> m_<instance_name> to declare a variable,
+and use an informative instance name.
+
+```systemverilog
+my_animal m_my_cat;
+my_animal m_my_dog;
+my_animal m_my_fish;
+```
+
+* If an array of class handles is desired, choose a name such that it follows
+one of the following schemes:
+
+```systemverilog
+int NUM_ANIMALS = 10;
+// This plural convention is recommended, but both are equally acceptable
+my_animal my_animals[NUM_ANIMALS];
+my_animal my_animal_array[NUM_ANIMALS];
+```
+
+
+To avoid name collisions, it is recommended that global types and types defined
+within a package have unique prefixes. This include typedefs, enumerated
+types/values, package parameters, package localparams, any functions/tasks, and
+macros.
+This prefixing is not necessary for any types defined within a class, as they
+will be scoped by the class.
+
+## Directory Structure and Packages
+
+*   Define one class per file, and the filename should match the name of the
+    class. The class files must be `` `include``-ed into the package file.
+*   Directory structure of all relevant testbench files should follow the
+example below:
+
+```systemverilog
+<block>_agent/
```

Comment:
Indicate that it is recommended to place the agent in a common area alongside other agents rather than with the DUT's UVM TB it was specifically written for - this creates a repository of UVC that users can find in one place.

Reply:
Good idea, I've added this point.